src/runtime/hexagon/hexagon/hexagon_buffer.cc

```diff
@@ -30,67 +34,135 @@ namespace tvm {
+
+template <HexagonBuffer::StorageScope S>
+std::unique_ptr<Allocation> Allocator(size_t nbytes, size_t alignment);
+
+template <>
+std::unique_ptr<Allocation> Allocator<HexagonBuffer::StorageScope::kDDR>(size_t nbytes,
+                                                                         size_t alignment) {
+  return std::make_unique<DDRAllocation>(nbytes, alignment);;
 }
 
-HexagonBuffer::HexagonBuffer(void* data, Optional<String> scope) : managed_{false} {
+template <>
+std::unique_ptr<Allocation> Allocator<HexagonBuffer::StorageScope::kVTCM>(size_t nbytes,
+                                                                          size_t alignment) {
+  return std::make_unique<VTCMAllocation>(nbytes, alignment);
+}
+
+HexagonBuffer::HexagonBuffer(size_t nbytes, size_t alignment, Optional<String> scope) : ndim_(1), nbytes_(nbytes) {
   SetStorageScope(scope);
-  allocations_.push_back(data);
+
+  std::unique_ptr<Allocation> alloca = nullptr;
+  if (GetStorageScope() == StorageScope::kDDR) {
+    alloca = Allocator<StorageScope::kDDR>(nbytes, alignment);
+  }
+  else if (GetStorageScope() == StorageScope::kVTCM) {
+    alloca = Allocator<StorageScope::kVTCM>(nbytes, alignment);
+  }
+  CHECK(alloca != nullptr);
+  allocations_.push_back(alloca->data_);
+  managed_allocations_.push_back(std::move(alloca));
 }
 
-HexagonBuffer::~HexagonBuffer() {
-  if (managed_) {
-    for (auto& ptr : allocations_) {
-      free(ptr);
+HexagonBuffer::HexagonBuffer(size_t ndim, size_t nbytes, size_t alignment, Optional<String> scope) : ndim_(ndim), nbytes_(ndim * nbytes) {
+  SetStorageScope(scope);
+  for (size_t i = 0; i < ndim; ++i) {
+    std::unique_ptr<Allocation> alloca = nullptr;
+    if (GetStorageScope() == StorageScope::kDDR) {
+      alloca = Allocator<StorageScope::kDDR>(nbytes, alignment);
+    }
+    else if (GetStorageScope() == StorageScope::kVTCM) {
+      alloca = Allocator<StorageScope::kVTCM>(nbytes, alignment);
     }
+    CHECK(alloca != nullptr);
+    allocations_.push_back(alloca->data_);
+    managed_allocations_.push_back(std::move(alloca));
   }
 }
 
-HexagonBuffer::HexagonBuffer(HexagonBuffer&& other)
-    : allocations_(other.allocations_),
-      managed_(other.managed_),
-      storage_scope_(other.storage_scope_) {
-  other.allocations_.clear();
-  other.managed_ = false;
-  other.storage_scope_ = StorageScope::kDDR;
+// TODO: add ExternalAllocation class?
+HexagonBuffer::HexagonBuffer(void* data, size_t nbytes, Optional<String> scope) : ndim_(1), nbytes_(nbytes) {
+  // TODO: disallow VTCM scope?
+  SetStorageScope(scope);
+  allocations_.push_back(data);
 }
 
-HexagonBuffer& HexagonBuffer::operator=(HexagonBuffer&& other) {
-  std::swap(allocations_, other.allocations_);
-  std::swap(managed_, other.managed_);
-  std::swap(storage_scope_, other.storage_scope_);
-  return *this;
+HexagonBuffer::~HexagonBuffer() {
+  managed_allocations_.clear();
 }
 
-void* HexagonBuffer::GetPointer() {
+void** HexagonBuffer::GetPointer() {
```

Comment:
Returning a void ** here disambiguates between the cases where there is a single allocation vs. multiple.  We always return a pointer to a pointer.  This may be controversial.

Reply:
Did we decide to revert this change so that tests pass and revisit once codegen supports pointer array indexing?

Reply:
I decided to keep `HexagonBuffer` as-is returning a `void**` which means that it (HexagonBuffer) supports discontiguous allocation.  And, I noted the portions of the device API that force / assume contiguous allocation awaiting RFC 39.